File: leet-code/206. Reverse Linked List/ReverseLinkedList206_2.java
public class ReverseLinkedList206_2 {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode next = reverseList(head.next);
        ListNode prev = new ListNode(head.val);
        ListNode cur = next;
        while (cur.next != null) {
            cur = cur.next;
        }
        cur.next = prev;
        return next;
    }
}

File: leet-code/206. Reverse Linked List/ReverseLinkedList206_2.kt
package com.example

class ReverseLinkedList206_2 {
    fun reverseList(head: ListNode?): ListNode? {
        return reverse(head, null)
    }

    private fun reverse(cur: ListNode?, prev: ListNode?): ListNode? {
        if (cur == null) return prev
        val next = cur.next
        cur.next = prev
        return reverse(next, cur)
    }
}